Spike/dsapi to sapint

bogdanadrianmarc opened the pull request:
Since this gem is currently only used by the LR apps, the changes are based around that. Originally, one of the apps would generate a query via the `QueryGenerator` class and that would be sent off to DSAPI. This PR changes that, so after the query is generated, before being sent off, the query is converted to SAPINT style and then sent off to a SAPINT instance. After that, the JSON in the response is converted to DSAPI JSON response style and returned to the LR app. Currently only the `explain` functionality is turned off, this will change after SAPINT implements that functionality

bogdanadrianmarc committed:
Added extra file to main gem file

bogdanadrianmarc committed:
SAPINT Converter class added

This handles converting a DSAPI query to a SAPINT one, and also changing the format of the returning JSON to make it look like it was returned from DSAPI

bogdanadrianmarc committed:
Changed default api url

bogdanadrianmarc committed:
Dataset endpoint modified

URL path is custom to LR, since this gem is only currently used by the LR apps; needs changing if it's going to be used with some other app

bogdanadrianmarc committed:
Query method now converts DSAPI to SAPINT

The method should convert the DSAPI to a SAPINT one, call the API with the latter, then convert the JSON response to a DSAPI one

bogdanadrianmarc committed:
Removed explain functionality

Currently waiting on a SAPINT explain functionality

ijdickinson reviewed (changes requested):
It would be much more idiomatic Ruby to define methods with `def` in the normal way, rather than have an array of lambdas. That would also allow you to factor out some of the common code (like dealing with `@id` hashes), and make it easier to call one method from another. To dispatch the conversions, Ruby's `case` statement would be a good solution.


In general, it would be better to create or transform arrays and then join them, rather than do string bashing. So instead of:

```ruby
      '@oneof' => lambda { |attribute, values|
        result = ''
        values.each do |value|
          result += "#{VALUE_FILTERS['@eq'].call(attribute, value)}&"
        end
        result.slice(0...-1)
      },
```

it would be better to do:

```ruby
      '@oneof' => lambda { |attribute, values|
        values.map do |value|
          VALUE_FILTERS['@eq'].call(attribute, value)
        end .join('&')
      },
```

The code to reformat the response from SapiNT to DSAPI format should be separated out, so that we can easily remove it if the view formatting can in future be done in SapiNT as I suggested.

We should really have tests for the new code

bogdanadrianmarc committed:
Refactored the converter class

This commit refactors the converter class and fixes a bug where SAPI would receive multiple searchPath params and return an error response

bogdanadrianmarc committed:
Moved the json converting part out of the query converter class

bogdanadrianmarc committed:
Added the query as Faraday params instead of already in URL

This is due to the fact that searching for multiple values on the same field with SAPI only rendered results based on the last value, because Faraday would only accept 1 value per field; now a list can be passed, which would be joined with '&' before sending the request

bogdanadrianmarc committed:
Explain functionality added

Explain now uses SAPI explain and works just like before

ijdickinson commented on lib/data_services_api/dataset.rb:
```diff
@@ -49,15 +49,54 @@ def structure
+    end
+
+    def to_dsapi_item(sapint_item)
+      dsapi_item = {}
```
[map](https://ruby-doc.org/core-3.0.1/Enumerable.html#method-i-map) is defined on `Enumerable`, so can be used directly on  a `Hash`:

```ruby
def to_dsapi_item(sapint_item)
  sapint_item.map { |k, v| [k, to_dsapi_json(k, v)] }.to_h
end
```

ijdickinson commented on lib/data_services_api/dataset.rb:
```diff
@@ -49,15 +49,54 @@ def structure
+    end
+
+    def to_dsapi_json(sapint_key, sapint_value)
+      dataset_name = data_api.split('/').slice(-1)
```
I think it would be clearer to use a regex with a capture group here

ijdickinson commented on lib/data_services_api/dataset.rb:
```diff
@@ -49,15 +49,54 @@ def structure
+      return { sapint_key => sapint_value } if sapint_key == '@id'
+      return { "#{dataset_name}:#{sapint_key}" => sapint_value } unless sapint_value.is_a?(Hash)
+
+      dsapi_json = {}
```
See note above about using `map` directly on a hash

ijdickinson commented on lib/data_services_api/dataset.rb:
```diff
@@ -49,15 +49,54 @@ def structure
+    private
+
+    # Converts SAPINT returned JSON format to DSAPI returned JSON format
+    def to_dsapi_response(sapint_response)
```
Converting to the DsAPI syntactic form is not a responsibility of a `Dataset` per se. I think it should be a separate class. That will also make it easier to write the tests

ijdickinson commented on lib/data_services_api/sapint_converter.rb:
```diff
@@ -0,0 +1,147 @@
+    end
+
+    # Converts a DSAPI query to SAPINT query
+    def to_sapint_query
```
I thought after our discussion on Thursday you were going to leave the URL parameters as a `Hash`, so that Faraday would handle them correctly? Or is `to_sapint_query` not used any more, in which case it should be deleted

ijdickinson commented on lib/data_services_api/sapint_converter.rb:
```diff
@@ -0,0 +1,147 @@
+
+    private
+
+    def sapint_query(key, value)
```
This is much clearer than the previous use of lambdas. :+1:

ijdickinson commented on lib/data_services_api/sapint_converter.rb:
```diff
@@ -0,0 +1,147 @@
+    end
+
+    def count(value)
+      '_count=@id' if value == true
```
You only need `if value` here

ijdickinson commented on lib/data_services_api/sapint_converter.rb:
```diff
@@ -0,0 +1,147 @@
+      "_offset=#{value}"
+    end
+
+    def andd(values)
```
The logic here is a bit hard to follow with the use of `value`, `values` and `[]` indexing. Can you please refactor to make the logic a bit clearer?

ijdickinson commented on lib/data_services_api/sapint_converter.rb:
```diff
@@ -0,0 +1,147 @@
+      end
+    end
+
+    def ge(attribute, value)
```
There seems to be an opportunity to DRY -up `ge`, `gt`, etc as they are all very similar

ijdickinson commented on lib/data_services_api/sapint_converter.rb:
```diff
@@ -0,0 +1,147 @@
+    def sanitize_search(search_text)
+      search_text
+        .gsub(' AND ', ' ')
+        .gsub('( ', '')
```
You can do both the parens in one step:

```
irb(main):005:0> q = "( foo AND bar )"
=> "( foo AND bar )"
irb(main):006:0> q.gsub(/(\( | \))/, '')
=> "foo AND bar"
irb(main):007:0> 
```

ijdickinson commented on lib/data_services_api/sapint_converter.rb:
```diff
@@ -0,0 +1,147 @@
+    end
+
+    def remove_prefix(value)
+      value.split(':')[1] unless value.slice(0...4) == 'http'
```
I think it would be clearer to say `unless value.match?(/^http/)`

ijdickinson commented on lib/data_services_api/sapint_converter.rb:
```diff
@@ -0,0 +1,147 @@
+    end
+
+    # Removes duplicate searchPaths and constructs JSON from params array
+    def sanitize_result(result)
```
If the URL params were left as a hash, you wouldn't need to split the string. As we discussed on Thursday, it's not a good idea to generate a string too early in the process

ijdickinson reviewed (commented):
I think the translation from the DsAPI structure to the SapiNT form is much clearer now. :+1: 

There are a few more comments to address, and I would like to see some reasonable amount of test coverage please.

bogdanadrianmarc committed:
Using map and reduce instead of each

bogdanadrianmarc committed:
Refactored response conversion to a separate class

bogdanadrianmarc committed:
Refactored converter and method

bogdanadrianmarc committed:
Drying up comparison methods in the converter

bogdanadrianmarc committed:
Constructing the SAPINT params as JSON now

bogdanadrianmarc commented on lib/data_services_api/service.rb:
```diff
@@ -3,7 +3,7 @@
   # Denotes the encapsulated DataServicesAPI service
   class Service
-    DEFAULT_URL = 'http://localhost:8080/dsapi'
+    DEFAULT_URL = 'http://localhost:8080'
```
The user can currently pass that in the `initialize` method for the `Service` class, do you mean to remove the default fallback value and always expect the API URL from the user? That would be set in each LR app container anyways, so I think it's ok to have a fallback value for development purposes at least

bogdanadrianmarc committed:
New service methods refactoring

bogdanadrianmarc committed:
Fixed old failing tests

This currently comments out the old failing tests and mentions that the feature is not currently available in SAPINT. They have not been completely removed because the tested methods and classes haven't been removed either

bogdanadrianmarc committed:
Added dsapi response converter tests

bogdanadrianmarc committed:
Fixed nil value bug for prefix removal

bogdanadrianmarc committed:
Added sapint converter tests

ijdickinson commented on spec/data_services_api/aspect_spec.rb:
```diff
@@ -1,53 +1,55 @@
-    _(@aspect1.service.url).wont_be_nil
-  end
-end
+# This feature is currently unavailable from SAPINT
```
I think it will be OK to just delete this spec. It will remain in the git history anyway

ijdickinson commented on spec/data_services_api/dataset_spec.rb:
```diff
@@ -12,29 +12,33 @@
-    end
-    # end
-  end
+  # This feature is currently unavailable from SAPINT
```
I would just delete these tests, and leave a comment to explain that to future readers

ijdickinson commented on lib/data_services_api/dsapi_response_converter.rb:
```diff
@@ -0,0 +1,36 @@
+      return { "#{dataset_name}:#{sapint_key}" => sapint_value } unless sapint_value.is_a?(Hash)
+
+      sapint_value.map do |key, value|
+        ["#{dataset_name}:#{sapint_key}#{key == '@id' ? '' : key.capitalize}", value]
```
When is the `key.capitalize` call used? I couldn't see any examples of this in the tests you've written

bogdanadrianmarc replied:
For example if we have:
```
...
propertyAddress: {
    ...
    county: ...
    ...
}
...
```
this should become:
```
...
propertyAddressCounty: ...
...
```
I'll add a test to try to simplify what's going on for a future reader 👍

ijdickinson reviewed (approved):
I think we are pretty close now, thanks. I've made a few minor suggestions, but nothing that blocks merging.

I **am** a bit concerned by the `santiize_search` call. Looking at the tests, it's clear that the "un-sanitized" is in Lucene syntax. I can't remember now why we needed to pass raw Lucene queries instead of search fragments, but we did have problems with Lucene not always giving us the results we wanted. We may have to switch to not-sanitising the search, and passing the strings as "raw" Lucene, which I believe that we can do with SapiNT. However, what I propose for now is that we merge these changes as-is, but then run some extensive tests on the apps when we have a new test environment up and running. The `hmlr-feature-tests` repo will also be helpful in this respect.

bogdanadrianmarc committed:
Removed api default url

This also updates the readme to reflect the change

bogdanadrianmarc committed:
Deleted unnecessary tests

There are some methods that remain in the gem but only work if a DSAPI URL is provided, so when we provide a SAPINT URL the tests for these methods fail; these tests were removed

bogdanadrianmarc committed:
Added extra test and small refactoring

bogdanadrianmarc committed:
Fixed failing tests due to default url removal
